tests/e2e/lib/apps.go

```diff
@@ -457,22 +457,156 @@ func makeRequest(request string, api string, todo string) {
+	if resp.StatusCode >= 200 && resp.StatusCode < 300 {
+		return serviceEndpoint, nil
+	}
+
+	return "", fmt.Errorf("Service endpoint not accessible (Status Code: %d)", resp.StatusCode)
+}
+
+// getApiEndpoint retrieves and verifies the accessibility of a Kubernetes service or route endpoint.
+//
+// Parameters:
+//
+//	clientv1:     An instance of the Kubernetes client.
+//	namespace:    The Kubernetes namespace in which the service or route is located.
+//	serviceName:  The name of the Kubernetes service (optional, routeName, serviceName or both must be provided).
+//	routeName:    The name of the Kubernetes service route (optional, routeName, serviceName or both must be provided).
+//
+// Returns:
+//
+//	string:       The full service or route endpoint URL if accessible.
+//	error:        An error message if both service and route endpoints are not accessible, if neither serviceName nor routeName is provided, or if there is an issue with the HTTP request.
+//
+// Example Usage:
+//
+//	apiEndpoint, err := getApiEndpoint(clientv1, "mongo-persistent", "todolist", todolist-route")
+//	if err != nil {
+//	    fmt.Println("Error:", err)
+//	} else {
+//	    fmt.Println("API endpoint is accessible at:", apiEndpoint)
+//	}
+func getApiEndpoint(clientv1 client.Client, namespace, serviceName, routeName string) (string, error) {
+	if serviceName == "" && routeName == "" {
+		return "", fmt.Errorf("Either serviceName or routeName must be provided")
+	}
+
+	if serviceName != "" {
+		serviceEndpoint, err := getServiceEndpoint(namespace, serviceName)
+		if err == nil {
+			return serviceEndpoint, nil
+		}
+		log.Printf("Failed to use Service endpoint for service %s in namespace %s: %v", serviceName, namespace, err)
+	}
+
+	if routeName != "" {
+		routeEndpoint, err := getRouteEndpoint(clientv1, namespace, routeName)
+		if err != nil {
+			return "", fmt.Errorf("Failed to get both service and route endpoints: %v", err)
+		}
+		return routeEndpoint, nil
+	}
+
+	return "", fmt.Errorf("serviceName or routeName must be provided")
+}
+
 // VerifyBackupRestoreData verifies if app ready before backup and after restore to compare data.
-func VerifyBackupRestoreData(clientv1 client.Client, artifact_dir string, namespace string, routeName string, app string, prebackupState bool, twoVol bool, backupRestoretype BackupRestoreType) error {
+func VerifyBackupRestoreData(clientv1 client.Client, artifact_dir string, namespace string, routeName string, serviceName string, app string, prebackupState bool, twoVol bool, backupRestoretype BackupRestoreType) error {
 	log.Printf("Verifying backup/restore data of %s", app)
-	appRoute := &routev1.Route{}
-	backupFile := artifact_dir + "/backup-data.txt"
-	routev1.AddToScheme(clientv1.Scheme())
-	err := clientv1.Get(context.Background(), client.ObjectKey{
-		Namespace: namespace,
-		Name:      routeName,
-	}, appRoute)
+
+	apiEndpoint, err := getApiEndpoint(clientv1, namespace, serviceName, routeName)
 	if err != nil {
-		return err
```

Comment:
I think this logic is incorrect, if error happens, it will try later to make request with empty `apiEndpoint`